modules/users/server/controllers/users.avatar.server.controller.js

```diff
@@ -0,0 +1,369 @@
+ */
+function getFacebookAvatarUrl(user, size) {
+  const id = _.get(user, ['additionalProvidersData', 'facebook', 'id'], false);
+
+  return (
+    id &&
+    `https://graph.facebook.com/${id}/picture/?width=${size}&height=${size}`
+  );
+}
+
+/**
+ * Generate url to avatar image that user uploaded
+ * @param {object} user - user object
+ * @param {number} size - size of the image
+ * @returns {string} - the url
+ */
+function getLocalAvatarUrl(user, size) {
+  const isValid = user && user.avatarUploaded && user._id;
+
+  if (isValid) {
+    // Cache buster
+    const timestamp = user.updated ? new Date(user.updated).getTime() : '';
+
+    // 32 is the smallest and 2048 biggest file size we're generating.
+    const fileSize = Math.min(Math.max(size, 32), 2048);
+
+    const domain = `${config.https ? 'https' : 'http'}://${config.domain}`;
+
+    return `${domain}/uploads-profile/${user._id}/avatar/${fileSize}.jpg?${timestamp}`;
+  }
+}
+
+/**
+ * Generate avatar url from Gravatar
+ * @link https://en.gravatar.com/site/implement/images/
+ * @param {object} user - user object
+ * @param {string} user.emailHash - gravatar identifies users by their email hashes
+ * @param {number} size - size of the image
+ * @returns {string} - the url
+ *
+ * @todo fallback image is provided from trustroots.org; it should rather come from config
+ */
+function getGravatarUrl(user, size) {
+  const isValid = user.emailHash;
+
+  // This fallback image has to be online one and not from localhost, since Gravatar needs to see it.
+  const fallbackImage = getDefaultAvatarUrl(size, false);
+
+  return (
+    isValid &&
+    `https://gravatar.com/avatar/${
+      user.emailHash
+    }?s=${size}&d=${encodeURIComponent(fallbackImage)}`
+  );
+}
+
+/**
+ * Generate avatar url
+ * @param {object} user - user object
+ * @param {integer} size - size of the image. Supported values are 2048, 1024, 512, 256, 128, 64, 36, 32, 24, 16.
+ * @param {string} source - avatar source. One of ['' (user's selected source), 'none', 'facebook', 'gravatar', 'local']
+ * @returns {string} - the url
+ */
+function getAvatarUrl(profile, size, source) {
+  return (
+    (source === 'local' && getLocalAvatarUrl(profile, size)) ||
+    (source === 'gravatar' && getGravatarUrl(profile, size)) ||
+    (source === 'facebook' && getFacebookAvatarUrl(profile, size)) ||
+    getDefaultAvatarUrl(size)
+  );
+}
+
+function getDefaultAvatarUrl(size = 1024, local = true) {
+  const domain = local
+    ? `${config.https ? 'https' : 'http'}://${config.domain}`
+    : 'https://trustroots.org';
+
+  return `${domain}/img/avatar-${size}.png`;
+}
+
+/**
+ * Serve avatar URL via redirect
+ *
+ * @TODO: serve by streaming instead of redirect and add caching layer.
+ *
+ * @param  {Object} res
+ * @param  {String} url
+ */
+function serveAvatarUrl(res, url) {
+  res
+    .status(302) // https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/302
+    // .setHeader('Cache-Control', 'public, max-age=0')
```

Comment:
@TODO: let's set some cache-time for these

Reply:
@rastapasta how does caching the image on mobile app work? Should we instruct it with this header from the API? That way we can change it remotely any time.